src/textual/widgets/_datatable.py

```diff
@@ -0,0 +1,317 @@
+                row_style = self.component_styles[component_row_style].node.rich_style
+            else:
+                row_style = base_style
+
+        scrollable_row = [
+            render_cell(row_index, column.index, row_style, column.width)[line_no]
+            for column in self.columns
+        ]
+
+        row_pair = (fixed_row, scrollable_row)
+        self._row_render_cache[cache_key] = row_pair
+        return row_pair
+
+    def _get_offsets(self, y: int) -> tuple[int, int]:
+        if self.show_header:
+            if y < self.header_height:
+                return (-1, y)
+            y -= self.header_height
+        return self._y_offsets[y]
+
+    def _render_line(
+        self, y: int, x1: int, x2: int, base_style: Style
+    ) -> list[Segment]:
+
+        width = self.content_region.width
+
+        cache_key = (y, x1, x2, width)
+        if cache_key in self._line_cache:
+            return self._line_cache[cache_key]
+
+        row_index, line_no = self._get_offsets(y)
+
+        fixed, scrollable = self._render_row(row_index, line_no, base_style)
+        fixed_width = sum(column.width for column in self.columns[: self.fixed_columns])
+
+        fixed_line: list[Segment] = list(chain.from_iterable(fixed)) if fixed else []
+        scrollable_line: list[Segment] = list(chain.from_iterable(scrollable))
+
+        segments = fixed_line + line_crop(scrollable_line, x1 + fixed_width, x2, width)
+
+        # line = Segment.adjust_line_length(segments, width, style=base_style)
+        remaining_width = width - (fixed_width + min(width, (x2 - x1 + fixed_width)))
+        if remaining_width > 0:
+            segments.append(Segment(" " * remaining_width, base_style))
+        elif remaining_width < 0:
+            segments = Segment.adjust_line_length(segments, width, style=base_style)
+
+        simplified_segments = list(Segment.simplify(segments))
+
+        self._line_cache[cache_key] = simplified_segments
+        return segments
+
+    @timer("render_lines")
+    def render_lines(self, crop: Region) -> Lines:
+
+        scroll_x, scroll_y = self.scroll_offset
+        x1, y1, x2, y2 = crop.translate(scroll_x, scroll_y).corners
+
+        base_style = self.rich_style
+
+        fixed_top_row_count = sum(
+            self.get_row_height(row_index) for row_index in range(self.fixed_rows)
+        )
+        if self.show_header:
+            fixed_top_row_count += self.get_row_height(-1)
+
+        render_line = self._render_line
+        fixed_lines = [
+            render_line(y, x1, x2, base_style) for y in range(0, fixed_top_row_count)
+        ]
+        lines = [render_line(y, x1, x2, base_style) for y in range(y1, y2)]
+
+        for line_index, y in enumerate(range(y1, y2)):
+            if y - scroll_y < fixed_top_row_count:
+                lines[line_index] = fixed_lines[line_index]
+
+        return lines
+
+    def on_mouse_move(self, event):
+        print(self.get_style_at(event.x, event.y).meta)
```

Comment:
this could probably be "decorated" with a `TODO` comment, as this is really not the kind of stuff we want to forget for the release? :sweat_smile:

Reply:
I won't forget. Still working on this class...